### PR TITLE
Msmq subscriptions are not persisted when transactional queues are in use

### PR DESCRIPTION
**Pull request by andreasohlund**
### Symptoms

Subscriptions are not stored when publishing endpoint stops.

### Who's affected

All Version 6 users using the [MSMQ subscription storag](https://docs.particular.net/nservicebus/msmq/subscription-persistence)e with [transactional queues](https://docs.particular.net/nservicebus/msmq/connection-strings).

### Details

This is a 6.0 bug only.

### CODE REVIEW

File: src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs [hunk 1]
```diff
@@ -51,6 +52,7 @@ public void TryReceiveById(string messageId)
         {
             try
             {
+                //todo: add failing test before fixing this
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs]: @timbussmann turns out that MessageQueueTransactionType.None works for both types of queues (see tests added). Should we leave as is with a comment?

Comment (danielmarbach) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs]: Would there be harm in changing and aligning it? Or is the intention of that question to keep the fix as focused as possible?

Comment (andreasohlund) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs]: Turned out this is a bug, this causes unsubscriptions to not be persisted. I'll write some tests and fix this

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers this one is ready for review

**Comment by andreasohlund** [comment]
Assigned to 6.0.2, I assume we don't want to bundle this with 6.0.1

File: src/NServiceBus.Core/Transports/Msmq/MsmqTransportInfrastructure.cs [hunk 2]
```diff
@@ -81,20 +81,18 @@ public override TransportReceiveInfrastructure ConfigureReceiveInfrastructure()
         {
             new CheckMachineNameForComplianceWithDtcLimitation().Check();
 
-            var builder = connectionString != null
-                ? new MsmqConnectionStringBuilder(connectionString).RetrieveSettings()
```
Comment (mikeminutillo) [inline src/NServiceBus.Core/Transports/Msmq/MsmqTransportInfrastructure.cs]: `ConfigureSendInfrastructure` later in this class makes this call again. Should it be updated to retrieve the one from the settings collection?

Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Msmq/MsmqTransportInfrastructure.cs]: Good catch @mikeminutillo , fixed in https://github.com/Particular/NServiceBus/pull/4343/commits/3808e042b2916a6a2a5eb3d3b3fd384eca0b3cda

File: src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs [hunk 3]
```diff
@@ -51,14 +52,15 @@ public void TryReceiveById(string messageId)
         {
             try
             {
-                q.ReceiveById(messageId, MessageQueueTransactionType.None);
+                queue.ReceiveById(messageId, MessageQueueTransactionType.None);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorageQueue.cs]: I think we talked about adding a comment here to make it clear that `None` is not a mistake but intention here?

File: src/NServiceBus.Msmq.AcceptanceTests/SubscriptionStorage/When_subscription_exists_for_non_tx_endpoint.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                queue.Send(new Message
+                {
+                    Label = Conventions.EndpointNamingConvention(typeof(Subscriber)),
+                    Body = typeof(MyEvent).AssemblyQualifiedName
+                }, MessageQueueTransactionType.None);
+            }
+
+            var ctx = await Scenario.Define<Context>()
+                .WithEndpoint<Publisher>(b =>
+                            b.When(c => c.EndpointsStarted, (session, c) => session.Publish(new MyEvent()))
```
Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/SubscriptionStorage/When_subscription_exists_for_non_tx_endpoint.cs]: formatting looks a bit off here

File: src/NServiceBus.Msmq.AcceptanceTests/SubscriptionStorage/When_subscription_exists_for_non_tx_endpoint.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                )
+                .WithEndpoint<Subscriber>()
+                .Done(c => c.GotTheEvent)
+                .Run(TimeSpan.FromSeconds(10));
```
Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/SubscriptionStorage/When_subscription_exists_for_non_tx_endpoint.cs]: why the short timeout?

Comment (andreasohlund) [inline src/NServiceBus.Msmq.AcceptanceTests/SubscriptionStorage/When_subscription_exists_for_non_tx_endpoint.cs]: All timeouts removed

**Comment by andreasohlund** [comment]
@timbussmann your comments are now addressed

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers this one is ready to go now

**Comment by andreasohlund** [comment]
Verified the fix using the pubsub sample and subscriptions are now persisted correctly

**Pull request merged**